openshift-kube-apiserver/filters/deprecatedapirequest/controller.go

```diff
@@ -0,0 +1,239 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers"
+)
+
+// NewController returns a Controller
+func NewController(node string, config *rest.Config) Controller {
+	return &controller{
+		client:       apiclientv1.NewForConfigOrDie(config).DeprecatedAPIRequests(),
+		nodeName:     node,
+		updatePeriod: 10 * time.Second,
+		incoming:     &requestCounts{Counts: map[string]map[int]map[string]map[string]int{}},
+		processing:   &requestCounts{Counts: map[string]map[int]map[string]map[string]int{}},
+	}
+}
+
+// Controller logs deprecated api requests
+type Controller interface {
+	Deprecated(resource, version, group string) bool
+	LogRequest(resource, version, group, verb, user string, timestamp time.Time)
+	Start(stop <-chan struct{})
+}
+
+// controller logs deprecated api requests periodically.
+type controller struct {
+	client       apiclientv1.DeprecatedAPIRequestInterface
+	nodeName     string
+	updatePeriod time.Duration
+	incoming     *requestCounts
+	processing   *requestCounts
+}
+
+// Deprecated return true if the resource is deprecated.
+func (c *controller) Deprecated(resource, version, group string) bool {
+	_, ok := deprecatedApiRemovedRelease[schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resource,
+	}]
+	return ok
+}
+
+// removedRelease of a specified resource.version.group.
+func (c *controller) removedRelease(r string) string {
+	s := strings.SplitN(r, ".", 3)
+	gvr := schema.GroupVersionResource{Resource: s[0], Version: s[1]}
+	if len(s) == 3 {
+		gvr.Group = s[2]
+	}
+	return deprecatedApiRemovedRelease[gvr]
+}
+
+// LogRequest queues an api request for logging
+func (c *controller) LogRequest(resource, version, group, verb, user string, ts time.Time) {
+	r := resource + "." + version
+	if len(group) > 0 {
+		r += "." + group
+	}
+	c.incoming.Lock()
+	defer c.incoming.Unlock()
+	c.incoming.IncrementRequestCount(r, verb, user, ts, 1)
+}
+
+// Start the controller
+func (c *controller) Start(stop <-chan struct{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-stop
+		klog.V(1).Infof("Shutting down deprecatedapirequest controller.")
+		cancel()
+	}()
+	go wait.Until(func() {
+		now := time.Now()
+		// clear old entries from processing
+		c.processing.trim(now)
+		// move incoming to processing
+		c.moveIncomingToProcessing()
+		// process processing queue
+		var wg sync.WaitGroup
+		for r := range c.processing.Counts {
+			go func(r string) {
+				wg.Add(1)
+				defer wg.Done()
+				c.processing.Lock()
+				counts := c.processing.Counts[r]
+				c.processing.Unlock()
+				err := c.ensureLogExists(ctx, r)
+				if err != nil {
+					runtime.HandleError(err)
+					return
+				}
+				_, _, err = v1helpers.UpdateStatus(ctx, c.client, r,
```

Comment:
naively, I would expect the flow to look like

1. copy current data from memory to a working location (this avoids a full lock)
2. update the API data for the requests in the last 24h
2. find the current hour from the API data and copy that data from the 24h index into the current hour in the API
3. on successful update, do nothing
4. on unsuccessful update, push the workingset back into "normal" memory location so we try again later.

That way we maintain a single source of truth and can easily update multiple indexes in the 24h list automatically